pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -1512,8 +1535,20 @@ void Object_beginTypeContext(JSOBJ _obj, JSONTypeContext *tc) {
         tc->type = JT_UTF8;
         return;
     } else if (object_is_decimal_type(obj)) {
-        GET_TC(tc)->doubleValue = PyFloat_AsDouble(obj);
-        tc->type = JT_DOUBLE;
+        /* Check for null, since null can't go thru double path */
+        PyObject *is_null_obj = PyObject_CallMethod(obj,
+                                                    "is_nan",
+                                                    NULL);
+        if (is_null_obj == Py_False) {
+            GET_TC(tc)->doubleValue = PyFloat_AsDouble(obj);
+            tc->type = JT_DOUBLE;
+        } else {
+            tc->type = JT_NULL;
+        }
+        if (!is_null_obj) {
```

Comment:
I think you should move this up to 1542 before the condition of equating to Py_False. This makes for a more consistent pattern where we always check for NULL after a `Py*` call